curator-recipes/src/main/java/org/apache/curator/framework/recipes/watch/InternalCuratorCache.java

```diff
@@ -0,0 +1,325 @@
+        {
+            @Override
+            protected void noteWatcherReset()
+            {
+                long count = refreshCount();
+                if ( (refreshOnStart && (count == 0)) || (count > 0) )
+                {
+                    internalRefresh(basePath, new NotifyingRefresher(InternalCuratorCache.this, basePath), cacheSelector);
+                }
+            }
+        };
+        watcher.getListenable().addListener(this);
+    }
+
+    @Override
+    protected void internalStart()
+    {
+        watcher.start();
+    }
+
+    @Override
+    protected void internalClose()
+    {
+        watcher.close();
+    }
+
+    @Override
+    public void process(WatchedEvent event)
+    {
+        switch ( event.getType() )
+        {
+            default:
+            {
+                // NOP
+                break;
+            }
+
+            case NodeDeleted:
+            {
+                remove(event.getPath());
+                break;
+            }
+
+            case NodeCreated:
+            case NodeDataChanged:
+            {
+                internalRefresh(event.getPath(), new Refresher(InternalCuratorCache.this), singleNodeCacheSelector);
+                break;
+            }
+        }
+    }
+
+    @Override
+    public CountDownLatch refreshAll()
+    {
+        return refresh(basePath);
+    }
+
+    @Override
+    public CountDownLatch refresh(String path)
+    {
+        if ( isStarted() && path.startsWith(basePath) )
+        {
+            CountDownLatch latch = new CountDownLatch(1);
+            Refresher refresher = new NotifyingRefresher(this, path, latch);
+            internalRefresh(path, refresher, cacheSelector);
+            return latch;
+        }
+        return new CountDownLatch(0);
+    }
+
+    @VisibleForTesting
+    volatile Exchanger<Object> debugRebuildTestExchanger;
+
+    private void internalRefresh(final String path, final Refresher refresher, final CacheSelector cacheSelector)
+    {
+        if ( !isStarted() )
+        {
+            return;
+        }
+
+        BackgroundCallback callback = new BackgroundCallback()
+        {
+            @Override
+            public void processResult(CuratorFramework client, CuratorEvent event) throws Exception
+            {
+                if ( event.getResultCode() == 0 )
+                {
+                    if ( event.getType() == CuratorEventType.GET_DATA )
+                    {
+                        CacheAction cacheAction = (CacheAction)event.getContext();
+                        CachedNode newNode = new CachedNode(event.getStat(), event.getData());
+                        CachedNode oldNode = putNewNode(path, cacheAction, newNode);
+                        if ( oldNode == null )
+                        {
+                            notifyListeners(CacheEvent.NODE_CREATED, path, newNode);
+                        }
+                        else if ( !nodeComparator.isSame(newNode, oldNode) )
```

Comment:
If cacheAction is `PATH_ONLY`, it'll return `nullNode`. Will this compare as no change, or always a change?  I'm thinking it should be always a change, because we don't know the old stat or data.